### ISSUE TITLE
実装: Frozen競り相手をconservative/passive 50:50混合にしてPASS崩壊を再検証

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#228]
Parent: #202
Depends on: #226
Base branch: `feature/issue-202-non-playing-ai`

## 背景

#226 で non-playing PPO rollout を candidate x1 + frozen x4 に変更し、candidate席を5席rotationする構成へ変更した。

その後、以下の条件で30 iterationsの短期再実験を実施した。

- run: `~/napoleon_runs/non-playing-ppo-frozen-v1`
- iterations: 30
- games-per-iteration: 200 logical seeds
- actual games per phase/iteration: 1000
- epochs: 4
- batch size: 128
- learning rate: 1e-4
- evaluation interval: 5
- evaluation games: 100 seeds x 5 rotations = 500 games
- seed: 202

実験自体は30/30 iterations完走し、最終evaluationは500/500 games完走、fallback=0 / illegal=0だった。

一方、bidding candidateのPASS率は以下のように増加した。

| iteration | PASS rate | candidate Napoleon formation rate | all-pass forced d=12 | mean reward |
|---:|---:|---:|---:|---:|
| 5 | 0.0467 | 0.963 | 30 | -0.347 |
| 10 | 0.1409 | 0.897 | 85 | 0.962 |
| 15 | 0.3529 | 0.721 | 188 | 2.219 |
| 20 | 0.7181 | 0.398 | 419 | 3.295 |
| 25 | 0.8522 | 0.269 | 498 | 3.405 |
| 30 | 0.9612 | 0.153 | 572 | 3.608 |

candidate x1 + frozen x4 によって以前より崩壊速度は遅くなったが、最終的には再びPASSへ収束している。

## 仮説

現在のFrozen bidding baselineは既存RuleBasedよりconservativeではあるものの、依然として競り参加率が高く、candidateがPASSしても他プレイヤーが高確率でNapoleonを引き受ける環境になっている可能性がある。

その場合、candidateは自分でNapoleonを引き受けるよりPASSして市民側へ回ることで高い期待rewardを得られ、PASS方策が合理的になっている可能性がある。

reward自体を変更する前に、Frozen bidding opponentの分布を広げてこの仮説を切り分ける。

## 目的

競り参加しにくい `passive` Frozen bidding baselineを追加し、bidding rolloutのFrozen4席を

- 50%: 現在の conservative frozen bidding baseline
- 50%: 新しい passive frozen bidding baseline

の混合にする。

candidateは従来どおり1席のみで、5席rotationする。

## passive bidding baseline

新しい学習modelは作らない。

既存 conservative / RuleBased bidding heuristicを基準に、パラメータ・閾値調整によって「さらに競り参加しにくい」固定方策を作る。

要件:

- conservative baselineより明確にPASS率が高い
- 完全PASS専用AIにはしない
- 強い手札ではNapoleonを狙う
- targetが高くなるほど競り上げ条件を厳しくする
- deterministic / seeded再現可能
- 学習中に更新されないFrozen policy
- 既存RuleBasedの大規模な再設計は行わない

具体的な閾値は既存実装を確認して合理的に調整する。

## 50:50混合方法

Frozen4席について、各席ごとに independently 50:50 で

- conservative
- passive

を選択する方式を基本とする。

ゲーム全体で4席すべてを同一baselineへ固定する方式にはしない。

これにより、candidateが「必ず誰かが競ってくれる」という単一の相手分布へ過適応しにくくする。

乱数選択はseedから決定論的に再現可能であること。

### bidding rollout

- learned bidding candidate x1
- frozen bidding opponents x4
  - conservative/passiveを各席50:50で決定論的に選択
- choosing-adjutant: Frozen固定
- exchange: Frozen固定
- playing: `ppo-separated-v1000` x5

### adjutant / exchange rollout

今回の主目的はbidding PASS崩壊の切り分けなので、不要な変更を入れない。
既存 #226 topologyを維持する。

## provenance / diagnostics

各rolloutについて、Frozen bidding opponent構成を追跡可能にする。

最低限:

- conservative/passiveそれぞれのpolicy identifier
- 各game / seatでどちらを使ったか再現可能な情報
- conservative seat count / passive seat count
- mixing rule/version

既存の以下のbidding diagnosticsは維持する。

- candidateBiddingDecisionCount
- passCount / passRate
- bidCount
- targetPointCardsDistribution 13〜19
- suitDistribution
- candidateNapoleonFormationRate
- declarationSuccessRate
- allPassForcedD12Count
- candidateRoleDistribution
- meanReward

## passive baseline固定seed評価

学習前に、同一の固定seed集合で少なくとも

1. 既存 RuleBased bidding
2. 現在の conservative frozen bidding
3. 新しい passive frozen bidding

を比較する。

passiveが conservativeより明確にPASS率が高いことを確認する。

ただし100% PASSにはしない。

## 30 iteration再実験

実装後、reward・PPO hyperparameter・candidate topologyは変更せず、今回のbaseline runと同条件で新規runを実施する。

推奨条件:

```text
iterations             30
games-per-iteration    200 logical seeds
epochs                  4
batch-size              128
learning-rate           1e-4
evaluation-interval     5
evaluation-games        100
seed                    202
```

旧run `~/napoleon_runs/non-playing-ppo-frozen-v1` は上書きしない。

## 比較対象

今回の既存baseline:

```text
iter30 PASS rate                 0.9612
iter30 candidate Napoleon rate  0.153
iter30 all-pass forced d=12     572
iter30 mean reward              3.608
fallback                        0
illegal                         0
```

新runについてiter 5/10/15/20/25/30の推移を同じ形式で比較する。

## Acceptance Criteria

- passive frozen bidding baselineが追加される
- passiveは固定seed評価でconservativeより明確にPASS率が高い
- passiveは完全PASS専用ではなく、十分強い手札ではbidする
- bidding rolloutのFrozen4席が各席50:50でconservative/passiveから選ばれる
- mixingはseeded deterministicで再現可能
- candidate x1 + frozen x4 + 5-seat rotationを維持する
- playingは全5席 `ppo-separated-v1000` frozen固定を維持する
- manifest / summaryからFrozen opponent mixを追跡できる
- 既存bidding diagnosticsを維持する
- 2〜3 iteration smokeが通る
- fallback=0 / illegal action=0
- 旧runを誤resumeしない
- READMEまたは実行手順を必要に応じて更新する
- 実装完了時に30 iteration比較実験用のコピペ可能なコマンドを提示する

## 実験結果の判定

少なくとも以下を見る。

- PASS率がiter30で再びほぼ100%へ張り付くか
- candidate Napoleon formation rateが低下し続けるか
- all-pass forced d=12が増え続けるか
- meanRewardとPASS率の正相関が弱まるか
- bid 13〜19分布が維持されるか

50:50混合でも同様にPASS崩壊する場合は、Frozen opponent分布だけでは説明できないと判断し、次にreward / all-pass forced contract構造の見直しへ進む。

## 非目標

- terminal role rewardの変更
- all-pass forced d=12ルールの変更
- PPOアルゴリズム変更
- learning rate等のhyperparameter tuning
- playing policy再学習
- opponent poolの高度化
- 親PR #203 のmainへのMerge

## PR運用

専用ブランチを最新 `feature/issue-202-non-playing-ai` から作成し、PR baseも `feature/issue-202-non-playing-ai` とする。

親PR #203 はmainへMergeしない。
